File: ruby/Dynamic Programming/rod_cutting.rb
module DynamicProgramming
  class RodCutting
    class << self

      # Internal: Method to calculate the maximum and optimal solution for cutting a
      #           rod of a particular length
      #           Recursive Strategy
      #
      # p - Price list to cut a rod of certain length
      # n - Length of rod to be cut
      #
      # Examples
      #   p = [1, 5, 8, 9, 10, 17, 17, 20, 24, 30]
      #   cut_rod(p, 9)
      #   => 31
      def cut_rod(p, n)
        return 0 if n == 0
        q = -Float::INFINITY
        (0..n-1).each { |i| q = [q, p[i] + cut_rod(p, n-i-1)].max }
        return q
      end

      # Internal: Method to calculate the maximum and optimal solution for cutting a
      #           rod of a particular length
      #           Recursive Memoization Technique
      #
      # p - Price list to cut a rod of certain length
      # n - Length of rod to be cut
      #
      # Examples
      #   p = [1, 5, 8, 9, 10, 17, 17, 20, 24, 30]
      #   memoized_cut_rod(p, 9)
      #   => 31
      def memoized_cut_rod(p, n)
        r = n.times.map { |x| -Float::INFINITY }
        memoized_cut_rod_aux(p, n, r)
      end


      def memoized_cut_rod_aux(p, n, r)
        return r[n-1] if r[n-1] >= 0
        if n == 0
          q = 0
        else
          q = -Float::INFINITY
          (0..n-1).each { |i| q = [q, p[i] + memoized_cut_rod_aux(p, n-i-1, r)].max }
        end
        r[n-1] = q
        q
      end

      def sleek_memoized_cut_rod_aux(p, n, r={})
        return 0 if n == 0
        q = -Float::INFINITY
        (0..n-1).each do |i|
          prevRodCut = r[n-i-1] || (r[n-i-1] = sleek_memoized_cut_rod_aux(p, n-i-1, r))
          q = [q, p[i] + prevRodCut].max
        end
        q
      end

      def bottom_up_cut_rod(p, n)
        r = n.times.map { |x| 0 }
        (1..n).each do |j|
          q = -Float::INFINITY
          (0..j-1).each { |i| q = [q, p[i] + r[j-i-1]].max }
          r[j] = q
        end
        r[n]
      end

      def extended_bottom_up_cut_rod(p, n)
        s = []
        s[0] = 0
        r = n.times.map{ |x| 0 }
        (1..n).each do |j|
          q = -Float::INFINITY
          (0..j-1).each do |i|
            if (q < p[i] + r[j-i-1])
              q = p[i] + r[j-i-1]
              s[j] = i+1
            end
            r[j] = q
          end
        end
        [r, s]
      end

      def print_cut_rod_solution(p, n)
        r, s = extended_bottom_up_cut_rod(p, n)
        while n > 0
          p s[n]
          n = n - s[n]
        end
      end
      
    end
  end
end


# p = [1, 5, 8, 9, 10, 17, 17, 20, 24, 30]
# p cut_rod(p, 9)
# p memoized_cut_rod(p, 9)
# p sleek_memoized_cut_rod_aux(p, 9)
# p bottom_up_cut_rod(p, 9)
# p extended_bottom_up_cut_rod(p, 10)

File: Java/DynamicProgramming/RodCutting.java
/**
 * Rod Cutting implementation in chapter 15 in the CLRS book.
 * It includes:
 * cutRod: Naive implementation using recursion. Time O(2^n).
 * memoizedCutRod: DP solution using top-down approach: memoize recursive algorithm. Time O(n^2).
 * bottomUpCutRod: DP solution using bottom-up approach: solve the subproblems in size order (smallest subproblems first). Time: O(n^2).
 * printCutRodSolution: prints the optimal solution with size of cuts.
 */

package dynamicprogramming;

public class RodCutting {

    public static void main(String[] args) {
        RodCutting rc = new RodCutting();
        int[] prices = { 1, 5, 8, 9, 10, 17, 17, 20, 24, 30 };

        for (int i = 0; i < prices.length + 1; i++) {
            int max0 = rc.cutRod(prices, i);
            int max1 = rc.memoizedCutRod(prices, i);
            int max2 = rc.bottomUpCutRod(prices, i);
            System.out.printf("recursive max: %d%n", max0);
            System.out.printf("memoized max: %d%n", max1);
            System.out.printf("bottomUp (dp) max: %d%n", max2);
            rc.printCutRodSolution(prices, i);
            System.out.println("----------------");
        }

    }
 
    //Naive version: recursive top-down implementation
    //Time: O(2^n)
    public int cutRod(int[] prices, int n) {
        if (n == 0) {
            return 0;
        }
        int max = Integer.MIN_VALUE;
        for (int i = 0; i < n; ++i) {
            max = Math.max(max, prices[i] + cutRod(prices, n - i - 1));
        }
        return max;
    }

    //Dynamic-programming: top-down approach with memoization
    //Time: O(n^2)
    public int memoizedCutRod(int[] prices, int n) {
        int[] revs = new int[n + 1];//revs[i] corresponds to the maximum revenues of length i. We define revs[0] = 0.
        for (int i = 0; i < revs.length; i++) {
            revs[i] = -1;//we use -1 here to indicate a state that the revs is not cached yet instead of negative infinity in the book because revenue is always nonnegative.
        }
        return memoizedCutRodAux(prices, n, revs);
    }

    private int memoizedCutRodAux(int[] prices, int n, int[] revs) {
        if (revs[n] >= 0) {
            return revs[n];
        }
        int max = Integer.MIN_VALUE;
        if (n == 0) {
            max = 0;
        } else {
            for (int i = 0; i < n; i++) {
                max = Math.max(max, prices[i] + memoizedCutRodAux(prices, n - i - 1, revs));
            }
        }
        revs[n] = max;
        return max;
    }
    
    //Dynamic-programming: bottom-up approach
    //Time: O(n^2)
    public int bottomUpCutRod(int[] prices, int n) {
        int[] revs = new int[n + 1];
        revs[0] = 0;//the revenue of a rod of length 0 is 0.
        int max = Integer.MIN_VALUE;
        for (int j = 1; j <= n; j++) {//revs[j] indicates maximum revenue of a rod of length j.
            max = Integer.MIN_VALUE;
            for (int i = 0; i < j; i++) {
                max = Math.max(max, prices[i] + revs[j - i - 1]);
            }
            revs[j] = max;
        }
        return revs[n];
    }


    public int[][] extendedBottomUpCutRod(int[] prices, int n) {
        int[] revs = new int[n + 1];
        int[] size = new int[n + 1];
        
        revs[0] = 0;
        int max = Integer.MIN_VALUE;
        for (int j = 1; j <= n; j++) {
            max = Integer.MIN_VALUE;
            for (int i = 0; i < j; i++) {
                if (max < prices[i] + revs[j - i - 1]) {
                    max = prices[i] + revs[j - i - 1];
                    size[j] = i + 1;
                }
            }
            revs[j] = max;
        }
        
        //For simplicity, return a 2d array where rs[0] is the revs array, rs[1] is the size array.
        //This may not be the optimized solution but I think it's cumbersome to create a tuple class in Java so I choose to use a 2D array.
        int[][] rs = new int[2][n + 1];
        for (int i = 0; i < n + 1; i++) {
            rs[0][i] = revs[i];
            rs[1][i] = size[i];
        }
        return rs;
    }

    public void printCutRodSolution(int[] prices, int n) {
        int[][] revsAndSize = extendedBottomUpCutRod(prices, n);
        int maxRevenue = revsAndSize[0][n];
        int[] size = revsAndSize[1];
        while (n > 0) {
            System.out.println(size[n]);
            n -= size[n];
        }
        System.out.printf("max revenue: %d%n", maxRevenue);
    }
}